py/farm_ng/core/events_file_reader.py

```diff
@@ -111,43 +127,56 @@ def num_frames(self, uri: Uri) -> int:
     def seek(self, uri: Uri, frame_id: int) -> None:
         assert uri.string() in self.offsets.keys()
         assert frame_id < self.num_frames(uri)
-        self._file_stream.seek(self.offsets[uri.string()][frame_id])
+        maybe_file_stream = self._file_stream
+        if maybe_file_stream is None:
+            return None
+        file_stream = cast(IO, maybe_file_stream)
+        file_stream.seek(self.offsets[uri.string()][frame_id])
 
     def seek_and_read(self, uri: Uri, frame_idx: int) -> Optional[Any]:
         self.seek(uri, frame_idx)
         return self.read()
 
-    def read(self) -> Optional[Any]:
+    def read(self) -> Optional[Tuple[Any, uri_pb2.Uri]]:
         # Read the message's length as bytes and convert it to an integer
-        event_len = int.from_bytes(self._file_stream.read(1), sys.byteorder)
+        maybe_file_stream = self._file_stream
+        if maybe_file_stream is None:
+            return None
+        file_stream = cast(IO, maybe_file_stream)
+
+        event_len = int.from_bytes(file_stream.read(1), sys.byteorder)
         if event_len == 0:
             self.close()
             return None
 
         # Read that number of bytes as the message bytes
-        event: bytes = self._file_stream.read(event_len)
+        event: bytes = file_stream.read(event_len)
 
         event_proto = event_pb2.Event()
         event_proto.ParseFromString(event)
 
         # parse the message
         message_cls = getattr(
-            importlib.import_module(event_proto.module), event_proto.name
+            importlib.import_module(event_proto.uri.scheme), event_proto.uri.authority
```

Comment:
I think we need to document and check how encode this in the uri. Ideally the protobuf type name (there is a descriptor in the protobuf) is used in the uri, and then this is used to look up the python package to import. In c++, a different mechanism will be used to automatically dlopen the shared library that has the protobuf definitions and then look up the c++ symbol name.